### PR TITLE
Add Windows AAD Pod Identity

### PR DESCRIPTION
**Pull request by bowu24**
**Reason for Change**:
This PR will add AAD Pod Identity supporting for Windows container

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

File: cmd/mic/main.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ var (
 )
 
 func main() {
-	klog.InitFlags(nil)
+	// klog.InitFlags(nil)
```
Comment (bhaskardeep) [inline cmd/mic/main.go:37]: Please undo this change #Resolved

File: go.mod [hunk 2]
```diff
@@ -1,37 +1,43 @@
 module github.com/Azure/aad-pod-identity
 
+// replace github.com/Azure/go-autorest v10.8.1+incompatible => github.com/Azure/go-autorest v0.3.0
```
Comment (bhaskardeep) [inline go.mod]: >v10 [](start = 40, length = 3)

remove #Resolved

File: cmd/mic/main.go (hunk 1 above)
Comment (bowu24) [reply cmd/mic/main.go:37]: If we uncomment this, mic will fail at this line for some reason. Need to do more investigation on this. #Resolved

File: pkg/nmi/server/redirector_windows.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	Server *Server
+}
+
+func makeRedirectorInt(server *Server) RedirectorInt {
+	return &WindowsRedirector{Server: server}
+}
+
+var podMap = make(map[types.UID]string)
+
+func (s *WindowsRedirector) RedirectMetadataEndpoint() {
```
Comment (bhaskardeep) [inline pkg/nmi/server/redirector_windows.go]: >RedirectMetadataEndpoint [](start = 28, length = 24)

Nit: please comments to public methods #Resolved

File: cmd/mic/main.go (hunk 1 above)
Comment (bowu24) [reply cmd/mic/main.go:37]: Synced with Anish and he suggests I can comment it for now. He will do more investigation.  #Resolved

File: go.mod (hunk 2 above)
Comment (bowu24) [reply go.mod]: Updated #Resolved

File: pkg/nmi/server/redirector_windows.go (hunk 3 above)
Comment (bowu24) [reply pkg/nmi/server/redirector_windows.go]: Updated #Resolved

**Comment by alukyan** [comment]
> 	}

Readability: This config is created in the middle of server initialization, when it does not seem to have any dependency on server. I suggest to move it up so that server initialization is not split in parts. #Resolved

---
Refers to: cmd/nmi/main.go:106 in 8e1764d. [](commit_id = 8e1764dc946395c4db8c32a33edb94ef31989820, deletion_comment = False)

File: pkg/nmi/server/server.go [hunk 4]
```diff
@@ -8,33 +8,44 @@ import (
...
 
+	aadpodid "github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity"
 	auth "github.com/Azure/aad-pod-identity/pkg/auth"
 	k8s "github.com/Azure/aad-pod-identity/pkg/k8s"
 	"github.com/Azure/aad-pod-identity/pkg/metrics"
 	"github.com/Azure/aad-pod-identity/pkg/nmi"
-	"github.com/Azure/aad-pod-identity/pkg/nmi/iptables"
 	"github.com/Azure/aad-pod-identity/pkg/pod"
 	"github.com/Azure/go-autorest/autorest/adal"
+	v1 "k8s.io/api/core/v1"
```
Comment (alukyan) [inline pkg/nmi/server/server.go:26]: >v1 "k8s.io/api/core/v1" [](start = 1, length = 23)

This package will have name "v1" no need to define same alias. Did you want to define `corev1`? #Resolved

Comment (alukyan) [reply pkg/nmi/server/server.go:26]: Oh, I see this is in preexisting code. hm.. feel free to ignore this.

---
In reply to: [383469572](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383469572) [](ancestors = 383469572)

File: pkg/nmi/server/server.go [hunk 5]
```diff
@@ -490,3 +460,13 @@ func validateResourceParamExists(resource string) bool {
 	}
 	return true
 }
+
+// Create the client config. Use kubeconfig if given, otherwise assume in-cluster.
+func buildConfig() (*rest.Config, error) {
+	kubeconfigPath := os.Getenv("KUBECONFIG")
```
Comment (alukyan) [inline pkg/nmi/server/server.go]: >KUBECONFIG [](start = 30, length = 10)

Why not add optional flag instead? #Resolved

File: pkg/pod/pod.go [hunk 6]
```diff
@@ -98,6 +116,58 @@ func (c *Client) GetPods() (pods []*v1.Pod, err error) {
...
+	}
+
+	return resList, nil
+}
+
+// GetPod returns the pod object
+func GetPod(obj interface{}, i informersv1.PodInformer) (pod *v1.Pod, err error) {
+
+	tries := 0
+	totalRetries := 300
```
Comment (alukyan) [inline pkg/pod/pod.go]: >totalRetries := 300 [](start = 1, length = 19)

We need to revise this logic - with 300 retries and explicit 1 second sleeps between them I am pretty sure we are using APIs in a wrong way. The fact that this call can take up to 5 minutes is definitely wrong way to use this API.

Comment (alukyan) [reply pkg/pod/pod.go]: AFAIK the "informer" pattern watches for object changes and notifies clients about those changes. Getting data store to which informer is connected and then starting explicitly polling for it looks like a bruteforce hack

---
In reply to: [383479968](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383479968) [](ancestors = 383479968)

File: pkg/nmi/server/redirector_windows.go (hunk 3 above)
Comment (alukyan) [reply pkg/nmi/server/redirector_windows.go]: Do you have Go linter in your VS code? Comments on public methods should follow certain format.

---
In reply to: [382734509](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r382734509) [](ancestors = 382734509)

File: pkg/nmi/server/server.go [hunk 7]
```diff
@@ -70,12 +96,14 @@ func NewServer(micNamespace string, blockInstanceMetadata bool) *Server {
...
+		PodObjChannel:         podObjCh,
 	}
 }
 
-// Run runs the specified Server.
-func (s *Server) Run() error {
-	go s.updateIPTableRules()
+// Run the specified Server.
+func (s *Server) Run(r RedirectorInt) error {
+	r.RedirectMetadataEndpoint()
```
Comment (alukyan) [inline pkg/nmi/server/server.go]: >r.RedirectMetadataEndpoint() [](start = 1, length = 28)

I do not see much use for this interface to be passed to a server - so it can call the redirector in the first line and forget about it.
It hides lot of important internal details about how redirector works and this does not help with maintaining the code in long run.

I would separate running redirector and running server into two independent parts. There is no reason to bunch them together.

Correctly designed interface will provide better answers to following questions:
* It is not clear how this redirector runs - is it one shot or a goroutine?
* It is not clear how redirector is parameterized - what parameters it depends on (needs)?
* If this redirector is a goroutine - when how it is terminated, is there any cleanup needed during termination?

Thoughts:
* Do we need a Redirector interface? in its current form a function type would work just as fine
* The pattern I would like to see is something like this in main.go:
```
// defined in server package
type RedirectorFunc func(ctx context.Context) // Receives a context which is Done() when redirector needs to stop

// in main

var redirector server.RedirectorFunc
if runtime.goos == "windows" {
    redirector = server.WindowsRedirector(....parameters needed for Windows ....)
} else {
    redirector = server.LinuxRedirector(.....parameters needed for Linux....)
}

// subscribe for shutdown signals and create context that gets canceled when shutdown signal is received
stopContext = ....
// call redirector explicitly in a goroutine

go redirector(stopContext)

// run server
server.Run()
```

Comment (alukyan) [reply pkg/nmi/server/server.go]: Platform specific implementations for `redirector_<platform>.go` should contain both methods (WindowsRedirector and LinuxRedirector) but one of them that is not for this platform should just panic

---
In reply to: [383488591](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383488591) [](ancestors = 383488591)

File: pkg/pod/pod.go (hunk 6 above)
Comment (bowu24) [reply pkg/pod/pod.go]: yes, I agree that it looks like a bruteforce hack. The reason for this is once the informer  detects a pod is created, it does not get the new pod's ip info: Pod.status.PodIP. Though wait for a while, the obj.(*v1.Pod) still does not return the Status.PodIP. So use  another call i.Informer().GetStore().Get(obj) to get its IP info. Let me say if I can so something like refresh to get ip info which can avoid the Getpod function here.

**Comment by alukyan** [comment]
> 	if err := http.ListenAndServe(":"+s.NMIPort, mux); err != nil {

I see NMI server does not do gracefull shutdown - is this OK?
https://golang.org/pkg/net/http/#Server.Shutdown #Resolved

---
Refers to: pkg/nmi/server/server.go:119 in 0b1cc1d. [](commit_id = 0b1cc1ddbc1e2b8d522a4189af087c9d6bcfd34c, deletion_comment = False)

File: cmd/nmi/main.go [hunk 8]
```diff
@@ -114,7 +114,10 @@ func main() {
 	if err = metrics.RegisterAndExport(*prometheusPort); err != nil {
 		klog.Fatalf("Could not register and export metrics: %+v", err)
 	}
-	if err := s.Run(); err != nil {
+
+	rd := s.GetRedirector()
+
+	if err := s.Run(rd); err != nil {
 		klog.Fatalf("%s", err)
 	}
```
Comment (alukyan) [inline cmd/nmi/main.go:150]: When run completes main will complete too and NMI will exit immediately.
It does not wait for redirector goroutine to perform graceful shutdown - which may leave IP tables not being cleaned up on Linux and routing not cleaned up on Windows.
This looks like a preexisting bug which needs to be fixed. #Resolved

**Review by alukyan — CHANGES_REQUESTED** [review]
:clock1:

File: pkg/nmi/server/endpointpolicy.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	if er != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteEndpointRoutePolicy applies the route policy against the pod ip endpoint
+func DeleteEndpointRoutePolicy(podIP string, metadataIP string) error {
+	if len(podIP) <= 0 {
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: ><= [](start = 15, length = 2)

I see you use it in more than one place - string cannot have negative length #Resolved

File: pkg/nmi/server/endpointpolicy.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		return errors.New("Missing IP Address")
+	}
+
+	endpoint, err := getEndpointByIP(podIP)
+	if err != nil {
+		return err
+	}
+
+	er := addEndpointPolicy(endpoint, metadataIP, metadataPort, nmiIP, nmiPort)
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: >er [](start = 1, length = 2)

This creates new `er` variable, but previous `err` variable is returned #Resolved

File: pkg/nmi/server/endpointpolicy.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+	klog.Infof("Getting endpoint for IP %s\n", ip)
+
+	request := msg.HNSRequest{
+		Entity:    msg.EndpointV1,
+		Operation: msg.Enumerate,
+		Request:   nil,
+	}
+
+	klog.Info("Enumerating all endpoints\n")
+	response, e := callHcnProxyAgent(request)
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: >e [](start = 11, length = 1)

Why do you use different error variable names throughout the code? `e, er, err` 
What each name means and why there are many? #Resolved

File: pkg/nmi/server/endpointpolicy.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+// DeleteEndpointRoutePolicy applies the route policy against the pod ip endpoint
+func DeleteEndpointRoutePolicy(podIP string, metadataIP string) error {
+	if len(podIP) <= 0 {
+		return errors.New("Missing IP Address")
+	}
+
+	endpoint, err := getEndpointByIP(podIP)
+	if err != nil {
+		return err
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: >err [](start = 9, length = 3)

I see you are not providing any context to the error - simply returning it back. Are you sure you can troubleshoot issues when you see error without a context? #Resolved

File: pkg/nmi/server/endpointpolicy.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+func addEndpointPolicy(endpoint *v1.HNSEndpoint, metadataIP string, metadataPort string, nmiIP string, nmiPort string) error {
+
+	if checkProxyPolicyExists(endpoint) == true {
+		klog.Infof("Proxy policy exists for endpoint %s. Skipping...\n", endpoint.Id)
+		return nil
+	}
+
+	klog.Infof("No proxy policy exists for the endpoint. Trying to apply policy to endpoint %s\n", endpoint.Id)
+	pp := &v1.ProxyPolicy{
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: >pp [](start = 1, length = 2)

policy
naming should be consistent throughout the code #Resolved

File: pkg/nmi/server/endpointpolicy.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+		return nil, e
+	}
+
+	var endpoints []v1.HNSEndpoint
+	err := json.Unmarshal(response, &endpoints)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, j := range endpoints {
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go]: >j [](start = 8, length = 1)

This is not loop variable, this is endpoint. `ep` would do for such small scope. `j` is just so confusing
Please pay attention to your naming practices #Resolved

File: cmd/mic/main.go (hunk 1 above)
Comment (bowu24) [reply cmd/mic/main.go:37]: Synced with Anish and we should comment this for now or mic will fail. He will do more investigation on this. 

---
In reply to: [381815036](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r381815036) [](ancestors = 381815036)

File: pkg/nmi/server/server.go (hunk 4 above)
Comment (bowu24) [reply pkg/nmi/server/server.go:26]: yes. I keep it consistent with the preexisting code.

---
In reply to: [383471881](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383471881) [](ancestors = 383471881,383469572)

File: pkg/nmi/server/endpointpolicy.go (hunk 9 above)
Comment (bowu24) [reply pkg/nmi/server/endpointpolicy.go]: Update it to be podIP == "". Let me know if you have any conceens, bhaskardeep.

---
In reply to: [383554501](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383554501) [](ancestors = 383554501)

**Comment by bowu24** [comment]
> 	if err := http.ListenAndServe(":"+s.NMIPort, mux); err != nil {

It looks like in preexisting code, it does not use the custom http server to control shutdown. In fct, the redirector goroute will try to accept the interrupt terminal signal and do the cleanup in for the IPTable or route policy.  For now, I think it is OK to leave it as preexisting code. I will try to find a better way to fix this graceful shutdown issue. 

---
In reply to: [590580313](https://github.com/bowu24/aad-pod-identity/pull/1#issuecomment-590580313) [](ancestors = 590580313)

---
Refers to: pkg/nmi/server/server.go:119 in 0b1cc1d. [](commit_id = 0b1cc1ddbc1e2b8d522a4189af087c9d6bcfd34c, deletion_comment = False)

File: pkg/nmi/server/redirector_windows.go (hunk 3 above)
Comment (bowu24) [reply pkg/nmi/server/redirector_windows.go]: yes. installed go outlinter. Will rewrite the comments

---
In reply to: [383483933](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383483933) [](ancestors = 383483933,382734509)

File: cmd/nmi/main.go (hunk 8 above)
Comment (bowu24) [reply cmd/nmi/main.go:150]: In redirector goroutine itself, it will try to wait for interrupt and terminal signal, then do the cleanup for ITTable and route policy. See handleTermination function in redirector_linux.go file. Do you think this will solve the problem here?

---
In reply to: [383552668](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383552668) [](ancestors = 383552668)

Comment (alukyan) [reply cmd/nmi/main.go:150]: I do not think that waiting in go routine will address this issue. The moment main() exits all goroutines will be terminated unconditionally. to solve the problem you need to make main() to wait for an event from goroutine before exiting - this can be done via closing a channel

---
In reply to: [386542574](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386542574) [](ancestors = 386542574,383552668)

Comment (bowu24) [reply cmd/nmi/main.go:150]: Got it. That makes sense. I implemented it by passing channel and closing chanel.

---
In reply to: [386636021](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386636021) [](ancestors = 386636021,386542574,383552668)

File: pkg/nmi/server/server.go (hunk 5 above)
Comment (bowu24) [reply pkg/nmi/server/server.go]: Could you explain a little more for this comment? Thanks.

---
In reply to: [383475214](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383475214) [](ancestors = 383475214)

File: cmd/mic/main.go (hunk 1 above)
Comment (alukyan) [reply cmd/mic/main.go:37]: This means that all `klog` settings will have their default values. I do not think this is desirable

---
In reply to: [386088105](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386088105) [](ancestors = 386088105,381815036)

File: pkg/nmi/server/server.go (hunk 5 above)
Comment (alukyan) [reply pkg/nmi/server/server.go]: you can define "kubeconfig" flag in nmi main()

Look at our deployment agent for example

---
In reply to: [386669543](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386669543) [](ancestors = 386669543,383475214)

Comment (bowu24) [reply pkg/nmi/server/server.go]: Got it. Updated.

---
In reply to: [386746627](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386746627) [](ancestors = 386746627,386669543,383475214)

File: pkg/nmi/server/redirector_windows.go [hunk 15]
```diff
@@ -16,7 +16,7 @@ import (
 var podMap = make(map[types.UID]string)
 
 // WindowsRedirector returns sync function for windows redirector
-func WindowsRedirector(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) func(*Server, chan bool, chan bool) {
+func WindowsRedirector(server *Server) func(*Server, chan bool, chan bool) {
 	exit := make(chan struct{})
```
Comment (alukyan) [inline pkg/nmi/server/redirector_windows.go]: This `exit` channel is completely private to the caller and is not connected to termination logic

File: pkg/nmi/server/endpointpolicy.go [hunk 16]
```diff
@@ -21,12 +21,12 @@ func ApplyEndpointRoutePolicy(podIP string, metadataIP string, metadataPort stri
 
 	endpoint, err := getEndpointByIP(podIP)
 	if err != nil {
-		return err
+		return fmt.Errorf("No endpoint found for Pod IP - %s. Error: %s", podIP, err.Error())
 	}
```
Comment (alukyan) [inline pkg/nmi/server/endpointpolicy.go:25]: newer version of Go support error wrapping where format specifier is %w and there is no need to use `err.Error()`.
Older version work very similar but without wrapping with %v specifier

File: pkg/nmi/server/redirector_linux.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+	// immediately.
+	updateIPTableRulesInternal(server)
+	server.Initialized = true
+
+	for {
+		select {
+		case <-mainRoutineDone:
+		case <-signalChan:
+			handleTermination()
+			subRoutinedone <- true
```
Comment (alukyan) [inline pkg/nmi/server/redirector_linux.go]: Correct semantics for stop channels is to close them. Not send a single message to them.
Closed channel will not block in multiple calls, but with single message the channel will block for the next message.
So if reader has multiple reads from the channel it will not handle termination correctly

File: pkg/nmi/server/redirector_linux.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	if err := iptables.LogCustomChain(); err != nil {
+		klog.Fatalf("%s", err)
+	}
+}
+
+// updateIPTableRules ensures the correct iptable rules are set
+// such that metadata requests are received by nmi assigned port
+// NOT originating from HostIP destined to metadata endpoint are
+// routed to NMI endpoint
+func updateIPTableRules(server *Server, subRoutinedone chan bool, mainRoutineDone chan bool) {
```
Comment (alukyan) [inline pkg/nmi/server/redirector_linux.go]: Channels could be Read, Write and R/W. Channel owner is the code that has Write permissions to the channel - it is the one who can close the channel. Common pattern in Go is to limit channel only to one of those modes which clearly communicates to the developers if this part of the code is supposed to only read, write to the channel or both. Passing channel with write permission also indicates the transfer of the ownership of the channel - the code that passes a write channel should not attempt to close it.

File: cmd/nmi/main.go (hunk 8 above)
Comment (bowu24) [reply cmd/nmi/main.go:150]: Update the code. Let me know if you have any comments.

File: cmd/nmi/main.go [hunk 19]
```diff
@@ -133,14 +132,13 @@ func main() {
...
 
-	go redirector(s, subRoutinedone, mainRoutineDone)
+	go redirector(s, mainRoutineDone)
 
 	if err := s.Run(); err != nil {
 		klog.Fatalf("%s", err)
 	}
 
 	close(mainRoutineDone)
```
Comment (alukyan) [inline cmd/nmi/main.go:142]: You basically reintroduced the issue that main function will not wait for the Sync() exit.
Please bring the channel back. You need to think about how to keep the ownership of the channel in redirector code and close the channel instead of sending single message.

Higher level issue is that top level functions do not have any unit and integration testing to catch issues like this. So unless someone mentions this in a code review it will go unnoticed

File: pkg/nmi/server/redirector_linux.go (hunk 18 above)
Comment (bowu24) [reply pkg/nmi/server/redirector_linux.go]: Add the directions to the channel.

Comment (bowu24) [reply pkg/nmi/server/redirector_linux.go]: Good point!

File: pkg/nmi/server/redirector_linux.go (hunk 17 above)
Comment (bowu24) [reply pkg/nmi/server/redirector_linux.go]: Good point.

File: pkg/nmi/server/server.go (hunk 7 above)
Comment (bowu24) [reply pkg/nmi/server/server.go]: Update the code as comments.

File: cmd/nmi/main.go (hunk 19 above)
Comment (bowu24) [reply cmd/nmi/main.go:142]: That is a bad push by accident... Reverted the change and add the direction for the channel.

File: pkg/nmi/server/endpointpolicy.go (hunk 16 above)
Comment (alukyan) [reply pkg/nmi/server/endpointpolicy.go:25]: Can you please address this comment about errors

File: pkg/nmi/server/redirector_windows.go (hunk 15 above)
Comment (alukyan) [reply pkg/nmi/server/redirector_windows.go]: what about this?

**Review by alukyan — CHANGES_REQUESTED** [review]

File: pkg/pod/pod.go (hunk 6 above)
Comment (alukyan) [reply pkg/pod/pod.go]: I see those Sleep() calls survived to the last iteration. These are potential time bombs that can cause all sorts of timing and performance issues in a busy working cluster.

**Review by alukyan — CHANGES_REQUESTED** [review]

File: pkg/nmi/server/redirector_windows.go (hunk 15 above)
Comment (bowu24) [reply pkg/nmi/server/redirector_windows.go]: Updated the code.

File: pkg/pod/pod.go (hunk 6 above)
Comment (bowu24) [reply pkg/pod/pod.go]: yes. You are absolutely right!!!!!! I never need to call the getpod function. Just check in the UpdateFunc is good enough. Will update the code. Thanks a lot for the great comments!

Comment (bowu24) [reply pkg/pod/pod.go]: I modify it by checking the ip change for the pods. Let me know if you have any comments.

File: cmd/nmi/main.go [hunk 20]
```diff
@@ -110,11 +123,30 @@ func main() {
...
+	mainRoutineDone := make(chan struct{})
+	subRoutineDone := make(chan struct{})
+
+	var redirector server.RedirectorFunc
+	if runtime.GOOS == "windows" {
+		redirector = server.WindowsRedirector(s, subRoutineDone)
+	} else {
+		redirector = server.LinuxRedirector(s, subRoutineDone)
 	}
+
```
Comment (alukyan) [inline cmd/nmi/main.go:135]: Looks like you are passing this channel twice to redirectors. First time when creating redirector, second time when you start redirector. Same for the server object.

Better pattern would be to create a channel in redirector constructor and return it back together with redirector. Return read-only channel - this indicates that main function cannot close it only redirector itself can do this. Closing `subroutineDone` channel in main will actually cause panic

As for server object it is better to capture it in constructor, that is returned redirector will already have an instance of `s`
```
       var redirectorDone <-chan struct{}
       if runtime.GOOS == "windows" {
		redirector, redirectorDone = server.WindowsRedirector(s)
	} else {
		redirector, redirectorDone = server.LinuxRedirector(s)
	}


       go redirector(mainDone)
```

File: pkg/pod/pod.go [hunk 21]
```diff
@@ -26,36 +25,58 @@ type Client struct {
...
 			},
 			UpdateFunc: func(OldObj, newObj interface{}) {
-				// We are only interested in updates to pod if the node changes.
+				oldPod := OldObj.(*v1.Pod)
+				newPod := newObj.(*v1.Pod)
+
+				// This is to handle windows nmi by observing ip change
+				if oldPod.Status.PodIP != newPod.Status.PodIP {
+					klog.V(6).Infof("Pod IP Updated")
+					klog.Infof("Old Pod IP: %s, Current Pod IP: %s", oldPod.Status.PodIP, newPod.Status.PodIP)
```
Comment (alukyan) [inline pkg/pod/pod.go:68]: Why have two logging lines side by side when we can have one. 
Also the top line has verbosity level V=6, line at the bottom has no verbosity level and produces log output always.
I suggest merging the lines and making logging verbosity consistent.

As for TODO: I suggest sweeping through the code we added/modified and making logging more consistent. If information we are logging is not important we can log it at higher verbosity level so we have cleaner logs, but we can enable more verbose logs if we need by just restarting container with different verbosity level setting
BTW: if you do not parse `klog` flags then logging verbosity will always be default.

This is non-blocking comment. Can be fixed in later iteration

**Review by alukyan — APPROVED** [review]
Looks good to me.
I left a couple of non-blocking comments that can be addressed in follow up PRs later.

File: cmd/nmi/main.go (hunk 20 above)
Comment (bowu24) [reply cmd/nmi/main.go:135]: As discussed in today's meeting, will file a task to track this issue and fix it in next iteration

File: pkg/pod/pod.go (hunk 21 above)
Comment (bowu24) [reply pkg/pod/pod.go:68]: This is for debug purpose. And I forgot to remove it. I will fix in next iteration.
